### PR TITLE
[NEP-491] Non-refundable storage 

### PR DESCRIPTION
**Pull request by jakmeier**
## Context
NEP: https://github.com/near/NEPs/pull/491
* Please note that action name was not agreed upon (see the end of the NEP discussion for more details). For now, I've used `NonrefundableStorageTransfer` name that was proposed in https://github.com/near/NEPs/pull/491#issuecomment-1876912603.

### Goal
We aim to enable the sponsorship of accounts with arbitrary storage (for example, with a contract that exceeds the zero balance account limit). Currently, there is a financial incentive to deplete such a sponsor by deleting the account and then cashing in.

## Summary
This PR introduces a non-refundable balance, which is a type of balance that can only be transferred to an account at the time of its creation. It is burned upon account deletion. This change eliminates the possibility of cashing out sponsored accounts.

### Changes
- Add `protocol_feature_nonrefundable_transfer_nep491` and change nightly protocol version.
- Introduce `AccountVersion::V2` as the new default version.
- Extend `Account` with a `nonrefundable` field that will only be used by accounts V2 or higher.
- Introduce `NonrefundableStorageTransfer` action that behaves similar to `Transfer` but it modifies the `nonrefundable`  instead of the `amount` account field. Both `nonrefundable` and `amount` are used to calculate the storage allowance.
- Introduce new error type `NonRefundableBalanceToExistingAccount` that it is raised on an attempt to make a non-refundable transfer outside of a transaction that creates an account.
- Use serialization sentinel (`u128::MAX`) as a first field during Borsh serialization in case of V2 accounts, to distinguish between V1 and V2 accounts. From V2, the `version` field is Borsh serialized and subsequent versions can use that field.
- Add custom Serde deserialization that uses V1 version in case the `version` field is missing. This is required when we parse the mainnet genesis file, so that the mainnet genesis hash does not change.
- Add non-refundable transfer tests.
- Add serde/borsh (de)serialization tests for all types of account.
- Update balance checker, `other_burnt_amount` is updated when deleting an account that had a non-refundable balance. Update related tests.
- Minor refactors.

### CODE REVIEW

**Comment by jakmeier** [comment]
This PR replaces https://github.com/near/nearcore/pull/9346 from a branch in nearcore, allowing @jbajic and others to contribute as well.

**Comment by akhi3030** [comment]
@jbajic: I will close this.  We can reopen if and when we decide to progress on this approach for the NEP.

**Comment by jbajic** [comment]
> @jbajic: I will close this. We can reopen if and when we decide to progress on this approach for the NEP.

I should have updated the PR, but I will be working on it, as was decided in the conversation that happened online. Posting the relevant comment here https://github.com/near/NEPs/pull/491#issuecomment-1790846143

File: integration-tests/src/tests/client/features/nonrefundable_transfer.rs
Comment (birchmd) [inline integration-tests/src/tests/client/features/nonrefundable_transfer.rs:115]: The number of boolean flags here all in a row make the tests a little hard to read in my opinion. Maybe it would be nicer (though more verbose) to pass a test config struct of some kind instead. Then at the callsite there would be something like

```rust
exec_transfer(env, signer, receiver, deposit, TransferConfig { nonrefundable: true, account_creation: false, ... })
```

File: integration-tests/src/tests/client/features/nonrefundable_transfer.rs
Comment (birchmd) [inline integration-tests/src/tests/client/features/nonrefundable_transfer.rs:226]: I think there should also be an assertion to check that the beneficiary account received the remaining balance from the deleted account, but none of the non-refundable balance.

File: integration-tests/src/tests/client/features/nonrefundable_transfer.rs
Comment (birchmd) [inline integration-tests/src/tests/client/features/nonrefundable_transfer.rs:290]: I think there should be a view call here to check the named account really does have the non-refundable balance it is supposed to have.

File: integration-tests/src/tests/client/features/nonrefundable_transfer.rs
Comment (birchmd) [inline integration-tests/src/tests/client/features/nonrefundable_transfer.rs:299]: I think there should be a view call here to check the implicit account really does have the non-refundable balance it is supposed to have.

File: integration-tests/src/tests/client/features/nonrefundable_transfer.rs
Comment (birchmd) [inline integration-tests/src/tests/client/features/nonrefundable_transfer.rs:308]: I think there should be a view call here to check the implicit account really does have the non-refundable balance it is supposed to have.

File: runtime/runtime/src/actions.rs
Comment (birchmd) [inline runtime/runtime/src/actions.rs:403]: Instead of having a boolean flag here it probably is better to have two separate functions: the original `action_transfer` and a new function `action_nonrefundable_transfer`. This better reflects the action model (there is a separate action now instead of a flag on the transfer action), and reduces complexity in the code by having less branching.

File: runtime/runtime/src/actions.rs
Comment (birchmd) [inline runtime/runtime/src/actions.rs:480]: Will it be possible to create an implicit account with both a normal balance an a non-refundable balance? I think it should be possible, and in that case there should be two separate `Balance` arguments instead of `Balance` and `bool`.

File: runtime/runtime/src/actions.rs
Comment (birchmd) [inline runtime/runtime/src/actions.rs:1066]: I think we need to revisit the `is_the_only_action` logic because now we have two kinds of transfers. Maybe this should be changed to `only_transfers`? Or Maybe there is no use-case for creating an implicit account with both normal and non-refundable balances in the one transaction, but then we will need to document this clearly.

File: runtime/runtime/src/balance_checker.rs
Comment (birchmd) [inline runtime/runtime/src/balance_checker.rs:527]: Hmm, this seems like a bad situation. Of course it can't really happen in production because no account will have that much $NEAR, but I still wonder if there's something we can do to avoid this side-effect of the sentinel-based distinction between account types.

**Review by birchmd — COMMENTED** [review]
Nice work! My main question is about if we want to allow implicit account creation with both normal and non-refundable balances. Otherwise mostly minor comments.

File: integration-tests/src/tests/client/features/nonrefundable_transfer.rs
Comment (staffik) [reply integration-tests/src/tests/client/features/nonrefundable_transfer.rs:290]: It is checked here:
https://github.com/near/nearcore/pull/9600/files#diff-14b73a0d63728329a6f04a38d1d2a7423bdd89edf2095debc5e94a8e48045f29R173

File: integration-tests/src/tests/client/features/nonrefundable_transfer.rs
Comment (staffik) [reply integration-tests/src/tests/client/features/nonrefundable_transfer.rs:299]: It is checked here:
https://github.com/near/nearcore/pull/9600/files#diff-14b73a0d63728329a6f04a38d1d2a7423bdd89edf2095debc5e94a8e48045f29R173

File: integration-tests/src/tests/client/features/nonrefundable_transfer.rs
Comment (staffik) [reply integration-tests/src/tests/client/features/nonrefundable_transfer.rs:308]: It is checked here:
https://github.com/near/nearcore/pull/9600/files#diff-14b73a0d63728329a6f04a38d1d2a7423bdd89edf2095debc5e94a8e48045f29R173

File: runtime/runtime/src/balance_checker.rs
Comment (staffik) [reply runtime/runtime/src/balance_checker.rs:527]: Will think about it. @jakmeier do you remember your thoughts on why this is the best (or the only viable) solution?

**Comment by walnut-the-cat** [comment]
@jakmeier and @birchmd , could you help review the latest PR commit?

**Comment by birchmd** [comment]
> @birchmd , could you help review the latest PR commit?

Looks like my question about whether or not we should allow implicit accounts with both regular and non-refundable balances remains unanswered. As well as my question about whether we want to use `Option<Balance>` instead of `Balance` to represent non-refundable storage in the account. These are both design questions. I'm not sure who the final decision maker is on those points.

File: integration-tests/src/tests/client/features/nonrefundable_transfer.rs
Comment (staffik) [reply integration-tests/src/tests/client/features/nonrefundable_transfer.rs]: I've modified tests so that 3 options are tested:
* `NonrefundableStorageTransfer` only
* both kind of transfers
  * non-refundable transfer first
  * regular transfer first

File: runtime/runtime/src/actions.rs
Comment (staffik) [reply runtime/runtime/src/actions.rs:478]: It is possible now, but for implicit account creation `NonrefundableStorageTransfer` must be the first action in the receipt. I could allow for regular transfer first, but this would require more invasive modification to the actions processing logic and we can always introduce this later.

File: runtime/runtime/src/balance_checker.rs
Comment (staffik) [reply runtime/runtime/src/balance_checker.rs:527]: We need to be able to distinguish between already existing V1 accounts and new V2 accounts during borsh deserialization. For that, we need a discriminator, which borsh serialized V1 accounts do not have.
If I understand it correctly, the only option to distinguish if the sequence of bytes we are deserializing is V1 or V2 account, is to encounter an invalid value for V1 account.
Possible choices are:
* `amount`
* `locked`
* `code_hash`
* `storage_usage`

`code_hash` cannot be used.
`amount` is better than `locked` because it is the first value.

I found
```
/// StorageUsage is used to count the amount of storage used by a contract.
pub type StorageUsage = u64;
```
If it really represents the amount of storage used by a contract then `amount` looks better, because it is `u128`.
Wdyt?

Comment (birchmd) [reply runtime/runtime/src/balance_checker.rs:527]: Yeah, I think you're right. This is probably the best solution and we'll leave it as is.

File: runtime/runtime/src/actions.rs
Comment (birchmd) [reply runtime/runtime/src/actions.rs:478]: Sure, requiring a specific order is ok I think. We just need to make sure it's documented somewhere.

**Review by birchmd — APPROVED** [review]

File: core/primitives-core/src/account.rs
Comment (staffik) [reply core/primitives-core/src/account.rs:338]: I would stick with what was approved in NEP.
There is the issue of constant `storage_num_bytes_account`: https://github.com/near/NEPs/pull/491#issuecomment-1710488417. And we can easily introduce this storage optimization later, thanks to having account version serialized from now.

Comment (birchmd) [reply core/primitives-core/src/account.rs:338]: Makes sense to me. Thanks!

File: runtime/runtime/src/lib.rs
Comment (jakmeier) [inline runtime/runtime/src/lib.rs:361]: Doesn't this new code accept transactions with multiple transfers to create implicit accounts? I think today a double transfer would fail if the account doesn't exist, yet.

If that is indeed what happens, you should use `checked_feature!` here instead of manual cfg.
Because during the protocol upgrade, some nodes will run the new code and some will use the old code, and we can't have them disagree on what is a valid account creation and what isn't.

But really, I think we should keep the behaviour the same. A refundable transfer followed by another transfer should keep failing unless there is a good reason to make that protocol change.

**Review by jakmeier — COMMENTED** [review]
@staffik Thank you for cleaning a lot of this up and making it all ready! It looks pretty good already.

I left several comments regarding changes I think you should consider. I think most things, if not all, could also be done in follow-up PRs. As long as everything is addressed (or dismissed) before the feature gets stabilized, it should be fine IMO.

But that's obviously your choice how you want to split the PRs. And a code owner will have to decide if it's really ready for merging to master.

File: core/primitives-core/src/account.rs
Comment (staffik) [reply core/primitives-core/src/account.rs:184]: Borsh-serialization output is hashed, but we need to parse mainnet genesis accounts as V1 because of that `nonrefundable: 0` that would come with V2.

File: core/primitives-core/src/account.rs
Comment (staffik) [reply core/primitives-core/src/account.rs:522]: We cannot really create new V1 accounts. It is possible for tests in this module, but for client these fields are private and it has to use`Account::new()`. `Account::new` always sets version to V2, and if nonrefundable feature is disabled it asserts that nonrefundable amount is 0.
So this test documents somewhat impossible situation if we already had an account V1 with nonrefundable amount greater than 0.

Comment (staffik) [reply core/primitives-core/src/account.rs:522]: I've just updated `Account::new()` so that it explicitly checks that nonrefundable is 0 if the default account version is V1.

File: core/primitives-core/src/account.rs
Comment (staffik) [reply core/primitives-core/src/account.rs:562]: As above, it is impossible situation, but it does not hurt to add the assertion you mentioned for serialization. I will update this.

File: runtime/runtime/src/lib.rs
Comment (staffik) [reply runtime/runtime/src/lib.rs]: For simplicity, I would just stick with the previous behavior, which is to require exactly one transfer action on implicit account creation. This was agreed in NEP and we can easily allow for multiple transfers later.
I would document that `NonrefundableStorageTransfer` has to be the only action in a receipt for implicit account creation.
cc @birchmd

Comment (birchmd) [reply runtime/runtime/src/lib.rs]: Sure, as long as it's documented I think either behaviour is fine.

Comment (jakmeier) [reply runtime/runtime/src/lib.rs]: Makes sense to me, too.

File: core/primitives-core/src/account.rs
Comment (jakmeier) [reply core/primitives-core/src/account.rs:524]: > If nonrefundable feature is enabled, `Account::new()` sets version to V2, otherwise it asserts that nonrefundable amount is 0.

Hm, I have to admit, I probably didn't think about this case properly when I wrote the reference implementation. But I believe to stabilize this feature, we must make account creation version dependent on the protocol version. Otherwise, validator nodes with the new binary will immediately start producing accounts of version V2, even if the protocol upgrade voting hasn't completed yet. That would cause different state roots for different nearcore binaries, and therefore validators would miss their chunks/blocks and get kicked out in the next epoch.

Does that make sense?

File: core/primitives-core/src/account.rs
Comment (jakmeier) [reply core/primitives-core/src/account.rs:193]: oh ok, I see :+1:

File: runtime/runtime/src/config.rs
Comment (staffik) [reply runtime/runtime/src/config.rs:108]: Added a TODO comment to consider this before stabilizing.

File: chain/rosetta-rpc/src/adapters/mod.rs
Comment (staffik) [reply chain/rosetta-rpc/src/adapters/mod.rs]: Added a TODO comment to consider this before stabilizing.

**Comment by staffik** [comment]
@nagisa Could you take a look at this as a code owner?

File: core/primitives-core/src/account.rs
Comment (nagisa) [inline core/primitives-core/src/account.rs:292]: Doesn't this allow `SENTINEL | 1 | AccountV2` to be successfully deserialized when `not(protocol_feature_nonrefundable_transfer_nep491)`? My thinking is that when this feature is disabled we should fail deserialization as we encounter the SENTINEL.

This also allows for multiple serialized forms that deserialize to the same thing.

File: core/primitives/src/test_utils.rs
Comment (nagisa) [inline core/primitives/src/test_utils.rs:686]: The name suggests the returned value would be gas, rather than NEAR. This wouldn't as big of a deal if Balance was a newtype, but with both it and Gas being plain aliases to regular integers (even if of different bitwidths) this can easily slip through.

Lets find at least a better name for this method.

File: integration-tests/src/tests/client/features/wallet_contract.rs
Comment (nagisa) [inline integration-tests/src/tests/client/features/wallet_contract.rs:67]: This change seems to be unrelated to the implementation of non-refundable storage. Would it be possible to split it out into its own NFC PR?

File: runtime/runtime/src/actions.rs
Comment (nagisa) [inline runtime/runtime/src/actions.rs:1021]: Minor nit: If this is written as

```
Action::Transfer(_) if account.is_none() => { ... }
Action::NonrefundableStorageTransfer(_) if account.is_none() => { ... }
```

Then upon eventual stabilization these branches can be merged together like many others elsewhere in this PR.

**Review by nagisa — COMMENTED** [review]
I wouldn't say I'm deeply familiar with any of the codebase areas touched by this PR, so this review is only a best effort outcome…

File: core/primitives-core/src/account.rs
Comment (staffik) [reply core/primitives-core/src/account.rs:185]: We do need `BorshDeserialize` at all, I've just removed it.
`LegacyAccount` is a helper struct that we only use to call `BorshSerialize` on it.

File: core/primitives-core/src/account.rs
Comment (staffik) [reply core/primitives-core/src/account.rs:292]: Yes, this branch is never entered if `not(protocol_feature_nonrefundable_transfer_nep491)`.
I've added an error return at the beginning of this branch.

The reason why we have
```
#[cfg(feature = "protocol_feature_nonrefundable_transfer_nep491")]
if version < AccountVersion::V2 {
```
is because `AccountVersion::V2` is introduced by `protocol_feature_nonrefundable_transfer_nep491`.

File: runtime/runtime/src/actions.rs
Comment (staffik) [reply runtime/runtime/src/actions.rs:1021]: We have `else` branch for `Action::NonrefundableStorageTransfer(_)`. I think we cannot merge these match arms.

File: chain/rosetta-rpc/src/lib.rs
Comment (staffik) [reply chain/rosetta-rpc/src/lib.rs:378]: Added TODO comment, will do as a follow up PR.

File: chain/rosetta-rpc/src/adapters/mod.rs
Comment (firatNEAR) [reply chain/rosetta-rpc/src/adapters/mod.rs]: I think this is a question that we should ask Coinbase since they own the spec. They usually are okay with making changes as long as they confirm from their side. @gmilescu could you help me with this, I am a bit out of loop with Coinbase.

File: runtime/runtime/src/config.rs
Comment (staffik) [reply runtime/runtime/src/config.rs]: @jakmeier Do you mean estimating cost for `NonrefundableStorageTransfer` action (which seems to be the same as for `Transfer` action IMO) or just introducing sth like `ActionCosts::nonrefundable_storage_transfer` with the same values as `ActionCosts::transfer` ?

File: runtime/runtime/src/actions.rs
Comment (nagisa) [reply runtime/runtime/src/actions.rs:1021]: Note the `if` guards in the example, so the overall post-merge state would look like this:

```rust
Action::Transfer(_) | Action::NonrefundableStorageTransfer(_) if account.is_none() => {
    /* the if branch */
}
Action::Transfer(_) => {}
Action::NonrefundableStorageTransfer(_) => { /* the else branch */ }
```

but either way a minor nit so fine to ignore it.

File: core/primitives-core/src/account.rs
Comment (nagisa) [reply core/primitives-core/src/account.rs]: Can you please walk me through why you claim this branch is never entered? My comment is operating under an assumption that anybody can handcraft some serialized data or is using fuzzer to generate inputs for this function, and that any reasoning based on what "trusted" serialization code does is invalid.

So my reading is that I can trivially get the code succeed the `sentinel_or_amount` check by starting my hand crafted payload with `0xFFFF_FFFF_FFFF_FFFF_FFFF_FFFF_FFFF_FFFF` and then there's nothing stopping me from setting the next field for `version_byte` to `1` and finishing with an `AccountV2` structure anyway.

The panic you have added suggests that this function never deals with any data that comes from potentially untrusted external sources, but that seems pretty difficult claim to verify. And it definitely needs to stand, as otherwise this panic here can have pretty severe consequences if a nefarious actor finds a way to execute it. In the worst case the entire network can be brought down for an extended time and do pretty nasty stuff during that downtime.

File: runtime/runtime/src/config.rs
Comment (jakmeier) [reply runtime/runtime/src/config.rs]: I was thinking about the second option. Using the same numerical value makes sense to me.
Creating a new parameter is mostly for consistency, which makes it easier to understand gas profiles and do all kinds of gas analysis work.

File: runtime/runtime/src/actions.rs
Comment (staffik) [reply runtime/runtime/src/actions.rs:1021]: I see, makes sense. Added TODO comment to merge it this way on stabilization.

File: core/primitives-core/src/account.rs
Comment (staffik) [reply core/primitives-core/src/account.rs]: Ah, I see now what you were up to.
Yes, the code potentially allowed to create V1 account using a sentinel.
I've changed it to return an error instead of panicking.
My previous comment was operating under an assumption that this branch is never entered under a `normal` operation in not-nightly and if there is a malicious input then one could set an arbitrary account `amount` anyway.
But yes, we have no guarantee how this code will be used so we must not panic there.

**Review by nagisa — COMMENTED** [review]
I do not see any further problems with the implementation code. One non-trivial operational point is that the coverage report is not favourable to this PR – it would be ideal to make sure the uncovered lines get tests added for them. I don't expect a full coverage, but I see entire match branches not being covered, which ain't great.

File: chain/rosetta-rpc/src/adapters/mod.rs
Comment (nasmithan) [reply chain/rosetta-rpc/src/adapters/mod.rs]: This change should be fine the way Rosetta handles this scenario. Thanks for checking.

**Review by nagisa — APPROVED** [review]

File: chain/rosetta-rpc/src/adapters/mod.rs
Comment (staffik) [reply chain/rosetta-rpc/src/adapters/mod.rs]: @nasmithan Could you take a look at this [change](https://github.com/near/nearcore/pull/10701/files#diff-843d19e05fcfdaab6b08c4cf1b67db9bf07a6805d60f2e2a9fdec45d85ff8a65L355-L382)?
In [discussion](https://github.com/near/NEPs/pull/491#issuecomment-1950280696) we decided to burn the nonrefundable balance instead of showing it as balance on the receiving account.

Comment (nasmithan) [reply chain/rosetta-rpc/src/adapters/mod.rs]: Yup that should be fine

**Pull request merged**
